lib/_http_outgoing.js

```diff
@@ -1005,6 +1026,131 @@ function connectionCorkNT(conn) {
   conn.uncork();
 }
 
+OutgoingMessage.prototype.writeFile = function writeFile(file, opts, callback) {
+  if (this[kLocked]) {
+    throw new ERR_STREAM_LOCKED('writeFile');
+  }
+
+  let start;
+  let end = Infinity;
+
+  if (typeof opts === 'function') {
+    callback = opts;
+    opts = null;
+  } else if (opts != null) {
+    validateObject(opts, 'opts');
+
+    start = opts.start;
+    if (start !== undefined) {
+      validateInteger(start, 'opts.start', 0);
+    }
+
+    end = opts.end;
+    if (end === undefined) {
+      end = Infinity;
+    } else if (end !== Infinity) {
+      validateInteger(end, 'opts.end', 0);
+
+      if (start !== undefined && start > end) {
+        throw new ERR_OUT_OF_RANGE('start', `<= "end" (here: ${end})`, start);
+      }
+    }
+
+    validateFunction(callback, 'callback');
+  }
+
+  try {
+    this[kLocked] = true;
+    if (typeof file === 'number') {
+      _writeFile(getValidatedFd(file), this, start, end, (err, bytesWritten) => {
+        this[kLocked] = false;
+        callback(err, bytesWritten);
+      });
+    } else if (typeof file === 'string' || isURL(file)) {
+      fs.open(getValidatedPath(file), (err, fd) => {
+        if (err) {
+          this[kLocked] = false;
+          callback(err);
+        } else {
+          _writeFile(fd, this, start, end, (err, bytesWritten) => {
+            fs.close(fd, (er) => {
+              this[kLocked] = false;
+              callback(er || err || null, bytesWritten);
+            });
+          });
+        }
+      });
+    } else {
+      // TODO (fix): Support FileHandle?
+      throw new ERR_INVALID_ARG_TYPE('file', ['string', 'number', 'URL'], file);
+    }
+  } catch (err) {
+    this[kLocked] = false;
+    throw err;
+  }
+};
+
+function _writeFile(fd, res, start, end, callback) {
```

Comment:
I think this function should be implemented in C++. There are too many JS -> C++ boundary calls.

Reply:
Follow up PR welcome 🙏

Reply:
There may be a challenge implementing this in c++ given how much of the implementation of `OutgoingMessage` is in JavaScript.